src/dbnode/storage/index/results.go

```diff
@@ -128,29 +120,29 @@ func (r *results) addDocumentsBatchWithLock(batch []doc.Metadata) error {
 	return nil
 }
 
-func (r *results) addDocumentWithLock(d doc.Metadata) (bool, int, error) {
-	if len(d.ID) == 0 {
-		return false, r.resultsMap.Len(), errUnableToAddResultMissingID
+func (r *results) addDocumentWithLock(w doc.Document) (bool, int, error) {
+	id, err := docs.ReadIDFromDocument(w)
+	if err != nil {
+		return false, r.resultsMap.Len(), err
 	}
 
-	// NB: can cast the []byte -> ident.ID to avoid an alloc
-	// before we're sure we need it.
-	tsID := ident.BytesID(d.ID)
+	if len(id) == 0 {
+		return false, r.resultsMap.Len(), errUnableToAddResultMissingID
+	}
 
 	// Need to apply filter if set first.
-	if r.opts.FilterID != nil && !r.opts.FilterID(tsID) {
+	if r.opts.FilterID != nil && !r.opts.FilterID(ident.BytesID(id)) {
```

Comment:
Would be great to avoid the `ident.BytesID(id)` allocation here (unfortunately we get an interface heap alloc...)

Reply:
Is this solution here to have a `FilterID` function that takes a byte slice?

Reply:
Possibly, but perhaps for now since this is using an exclusive lock, can we use a `*ident.ReuseableBytesID` here that is used with:
```go
r.reuseableID.Reset(id)
if r.opts.FilterID != nil && !r.opts.FilterID(r.reuseableID) {
  return false, r.resultsMap.Len(), nil
}
```